File: examples/contract_multicall_async/contract_multicall_async.vb
Module contract_call_async

    Sub Main()

        Using client As New AergoClient()

            client.Connect("testnet-api.aergo.io", 7845)

            Dim account As AergoClient.AergoAccount = New AergoClient.AergoAccount With
                {.privkey = New Byte() {&HDB, &H85, &HDD, &HC, &HBA, &H47, &H32, &HA1, &H1A, &HEB,
                                        &H3C, &H7C, &H48, &H91, &HFB, &HD2, &HFE, &HC4, &H5F, &HC7,
                                        &H2D, &HB3, &H3F, &HB6, &H1F, &H31, &HEB, &H57, &HE7, &H24,
                                        &H61, &H76}}

            ' or use an account on Ledger Nano S
            'Dim account As AergoClient.AergoAccount = New AergoClient.AergoAccount With
            '    {.use_ledger = True}

            Console.WriteLine("Calling Smart Contract Function Asynchronously...")

            Dim script As String = "[" & _
                "[""call"",""AmhcceopRiU7r3Gwy5tmtkk4Z3Px53SfsKBifGMvaSSNiyWrvKYe"",""set_name"",""multicall async""]," & _
                "[""call"",""AmhQebt5N4pikoE4fZdEbaRAExoTwrLG3F5pKoGBVRbG5KeNKeoi"",""call_func""]," & _
                "[""call"",""AmgqG45bTohBSQUgfDK4z31fq5jTy7CeRdwAf46hfaATG6tZtShf"",""set"",""1"",""first""]," & _
                "[""call"",""AmgqG45bTohBSQUgfDK4z31fq5jTy7CeRdwAf46hfaATG6tZtShf"",""set"",""2"",""second""]," & _
                "[""call"",""AmgqG45bTohBSQUgfDK4z31fq5jTy7CeRdwAf46hfaATG6tZtShf"",""get"",""1""]," & _
                "[""return"", ""%last result%""]" & _
            "]"

            Console.WriteLine(script)

            Dim context = New AergoClient.CallbackState() With {.data = "you can pass any object to the callback here"}

            Dim ret = client.MultiCallAsync(AddressOf CallBack, context, account, script)

            If ret Then
                Console.WriteLine("Waiting for response...")
                While (client.ProcessRequests(5000) > 0)
                    ' loop
                End While
            Else
                Console.WriteLine("Failed to send the transaction")
            End If

        End Using

        Console.WriteLine("Press any key to continue")
        Console.ReadKey()
    End Sub

    Private Sub CallBack(ByVal context As AergoClient.CallbackState, ByVal receipt As AergoClient.TransactionReceipt)

        Console.WriteLine("--- Transaction Receipt ---")
        Console.WriteLine("Status : " + receipt.status)
        Console.WriteLine("ret    : " + receipt.ret)
        Console.WriteLine("BlockNo: " + receipt.blockNo.ToString())
        Console.WriteLine("TxIndex: " + receipt.txIndex.ToString())
        Console.WriteLine("GasUsed: " + receipt.gasUsed.ToString())

    End Sub

End Module

File: examples/contract_multicall_async/contract_multicall_async.cs
using System;
using System.Text;

namespace contract_call_async_cs
{
    class contract_call_async
    {

        public static void ReceiptReceived(AergoClient.CallbackState context, AergoClient.TransactionReceipt receipt)
        {
            Console.WriteLine("--- Transaction Receipt ---");
            Console.WriteLine("Status : " + receipt.status);
            Console.WriteLine("ret    : " + receipt.ret);
            Console.WriteLine("BlockNo: " + receipt.blockNo);
            Console.WriteLine("TxIndex: " + receipt.txIndex);
            Console.WriteLine("GasUsed: " + receipt.gasUsed);
        }

        static void Main(string[] args)
        {
            using (AergoClient client = new AergoClient())
            {
                client.Connect("testnet-api.aergo.io", 7845);

                AergoClient.AergoAccount account = new AergoClient.AergoAccount
                {
                    privkey = new byte[] {
                        0xDB, 0x85, 0xDD, 0x0C, 0xBA, 0x47, 0x32, 0xA1, 0x1A, 0xEB,
                        0x3C, 0x7C, 0x48, 0x91, 0xFB, 0xD2, 0xFE, 0xC4, 0x5F, 0xC7,
                        0x2D, 0xB3, 0x3F, 0xB6, 0x1F, 0x31, 0xEB, 0x57, 0xE7, 0x24,
                        0x61, 0x76 }
                };

                /* or use an account on Ledger Nano S */
                //AergoClient.AergoAccount account = new AergoClient.AergoAccount
                //{
                //    use_ledger = true
                //};

                Console.WriteLine("Calling Smart Contract Function Asynchronously...");

                string script = @"[
                    [""call"",""AmhcceopRiU7r3Gwy5tmtkk4Z3Px53SfsKBifGMvaSSNiyWrvKYe"",""set_name"",""multicall async""],
                    [""call"",""AmhQebt5N4pikoE4fZdEbaRAExoTwrLG3F5pKoGBVRbG5KeNKeoi"",""call_func""],
                    [""call"",""AmgqG45bTohBSQUgfDK4z31fq5jTy7CeRdwAf46hfaATG6tZtShf"",""set"",""1"",""first""],
                    [""call"",""AmgqG45bTohBSQUgfDK4z31fq5jTy7CeRdwAf46hfaATG6tZtShf"",""set"",""2"",""second""],
                    [""call"",""AmgqG45bTohBSQUgfDK4z31fq5jTy7CeRdwAf46hfaATG6tZtShf"",""get"",""1""],
                    [""return"", ""%last result%""]
                ]";

                Console.WriteLine(script);

                var context = new AergoClient.CallbackState() { data = "any object can be set here as context" };

                var ret = client.MultiCallAsync(ReceiptReceived, context, ref account, script);

                if (ret) {
                    Console.WriteLine("Waiting for response..");
                    while (client.ProcessRequests(5000) > 0) {
                        // loop
                    }
                } else {
                    Console.WriteLine("Failed to send the transaction");
                }

                Console.WriteLine("Press any key to continue");
                Console.ReadKey();
            }
        }
    }
}
